# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ayyanatariq03-arch/Flyrank-ML/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [9]:
import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/ayyanatariq03-arch/Flyrank-ML"
REPO_DIR = "Flyrank-ML"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)

print("Working dir:", os.getcwd())
assert os.path.exists("data/raw/content_refresh_anonymized.csv"), "CSV not found — check repo structure"
print("Data found. Ready.")

Working dir: /content/Flyrank-ML/Flyrank-ML
Data found. Ready.


## 1. My lane (or freestyle) and why

*## My Lane: Refresh / Content Opportunity Scoring

I'm choosing Lane 2 — Refresh / Content Opportunity Scoring. In Notebooks 01
and 02, I already saw that a simple hand-written rule (stale + visible pages)
gets decent precision, but a learned model can sometimes beat it, especially
deeper in a ranked list. This lane is a natural extension: build a ranked
queue that tells a content team which pages to review first, with reason
codes they can trust. Content teams don't have time to manually check every
page, so a well-prioritized queue is a genuinely useful, usable output.*

In [10]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

**Decision:** Which pages should the content/SEO team review first for
refresh, expansion, or protection?

**Who acts on it:** Content managers or SEO specialists with limited weekly
review capacity.

**Action:** They pull the top N pages from a ranked refresh queue (with
reason codes) and decide whether to update, expand, or leave a page.

**Cost of a wrong recommendation:** A false positive wastes limited editor
time reviewing a page that wasn't actually at risk. A false negative means a
genuinely declining page keeps losing visibility unnoticed until the damage
is harder to reverse. Both costs are opportunity costs — time and visibility.*

In [11]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [12]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

total_pages = df.shape[0]
declining_rate = df["trend_direction"].str.lower().eq("down").mean()
avg_impressions_declining = df[df["trend_direction"].str.lower() == "down"]["impressions_90d"].mean()

print(f"Total pages: {total_pages:,}")
print(f"Share currently declining: {declining_rate:.1%}")
print(f"Avg impressions_90d for declining pages: {avg_impressions_declining:.0f}")


Total pages: 30,000
Share currently declining: 54.2%
Avg impressions_90d for declining pages: 4919


## Quick look at the data

From the starter dataset: 30,000 pages total, and 54.2% are currently
declining — more than half the dataset, which makes this a real, sizeable
prioritization problem rather than a rare edge case. Declining pages still
average [AVG_IMPRESSIONS] impressions in the last 90 days, meaning they
haven't lost all visibility yet — there's still traffic worth protecting if
caught early.

In the earlier pipeline run (Notebooks 01/02), a random forest model reached
Precision@50 of 0.740 versus the baseline rule's 0.240 — roughly 3x better at
flagging real decliners in the top 50. That gap is exactly what this lane is
built to close and act on.


## 4. Careful words: what I can and can't claim

***What I can say:** This project produces an observed, directional ranked
queue — decision-support for a human reviewer, based on the last 90 days of
history for the same pages.

**What I will never claim:** I cannot claim a refresh will cause a recovery
(that needs a real experiment), and I am not predicting Google's algorithm —
only finding statistical patterns in historical, observable data that may not
hold for new pages or future periods.*

In [13]:
print("Claims: observed, directional, decision-support only")
print("Not claiming: causal proof or predicting Google's algorithm")

Claims: observed, directional, decision-support only
Not claiming: causal proof or predicting Google's algorithm


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.